pkg/metrics/sketch_series.go

```diff
@@ -106,98 +107,198 @@ func (sl SketchSeriesList) MarshalJSON() ([]byte, error) {
 // it's contents are marshaled individually, packed with the appropriate protobuf metadata, and compressed in stream.
 // The resulting payloads (when decompressed) are binary equal to the result of marshaling the whole object at once.
 func (sl SketchSeriesList) MarshalSplitCompress(bufferContext *marshaler.BufferContext) ([]*[]byte, error) {
-	// The Metadata field of gogen.SketchPayload is never written to - so pack an empty metadata as the footer
-	footer := []byte{0x12, 0}
+	var err error
+	var compressor *stream.Compressor
+	buf := bufferContext.PrecompressionBuf
+	ps := molecule.NewProtoStream(buf)
+	payloads := []*[]byte{}
+
+	// constants for the protobuf data we will be writing, taken from
+	// https://github.com/DataDog/agent-payload/blob/a2cd634bc9c088865b75c6410335270e6d780416/proto/metrics/agent_payload.proto#L47-L81
+	const payloadSketches = 1
+	const payloadMetadata = 2
+	const sketchMetric = 1
+	const sketchHost = 2
+	const sketchDistributions = 3
+	const sketchTags = 4
+	const sketchDogsketches = 7
+	/* unused
+	const distributionTs = 1
+	const distributionCnt = 2
+	const distributionMin = 3
+	const distributionMax = 4
+	const distributionAvg = 5
+	const distributionSum = 6
+	const distributionV = 7
+	const distributionG = 8
+	const distributionDelta = 9
+	const distributionBuf = 10
+	*/
+	const dogsketchTs = 1
+	const dogsketchCnt = 2
+	const dogsketchMin = 3
+	const dogsketchMax = 4
+	const dogsketchAvg = 5
+	const dogsketchSum = 6
+	const dogsketchK = 7
+	const dogsketchN = 8
+
+	// generate a footer containing an empty Metadata field (TODO: this isn't
+	// necessary; an omitted field will be assumed empty)
+	var footer []byte
+	{
+		buf := bytes.NewBuffer([]byte{})
+		ps := molecule.NewProtoStream(buf)
+		ps.Embedded(payloadMetadata, func(ps *molecule.ProtoStream) error {
+			return nil
+		})
+		footer = buf.Bytes()
+	}
+
+	// Prepare to write the next payload
+	startPayload := func() error {
+		var err error
 
-	bufferContext.CompressorInput.Reset()
-	bufferContext.CompressorOutput.Reset()
+		bufferContext.CompressorInput.Reset()
+		bufferContext.CompressorOutput.Reset()
 
-	compressor, e := stream.NewCompressor(bufferContext.CompressorInput, bufferContext.CompressorOutput, []byte{}, footer, []byte{})
-	if e != nil {
-		return nil, e
+		compressor, err = stream.NewCompressor(bufferContext.CompressorInput, bufferContext.CompressorOutput, []byte{}, footer, []byte{})
+		if err != nil {
+			return err
+		}
+
+		return nil
 	}
-	payloads := []*[]byte{}
 
-	dsl := make([]gogen.SketchPayload_Sketch_Dogsketch, 1)
-	for _, ss := range sl {
-		if len(ss.Points) > cap(dsl) {
-			dsl = make([]gogen.SketchPayload_Sketch_Dogsketch, len(ss.Points))
+	finishPayload := func() error {
+		var payload []byte
+		// Since the compression buffer is full - flush it and rotate
```

Comment:
This comment used to be in the switch under `errPayloadFull` - perhapse it should be moved back as I think it contextually makes more sense there.